### DISCUSSION TITLE
Custom loss function

### DISCUSSION THREAD

**Discussion by rnjv** [Nixtla/neuralforecast#575] [General]
Hi,

I appreciate if anyone can provide guidance on implementing a custom loss function.

I am looking at https://github.com/marcdemers/batch-DILATE/tree/main and https://github.com/vincent-leguen/DILATE and it looks interesting.

Thank you!

**Comment by kdgutier** [comment]
Hey @rnjv,

If you have more questions and reach us in the [slack channel](https://nixtlacommunity.slack.com/ssb/redirect).

You need to create a Pytorch module with the following methods: 
(i) `__init__` modified DILATE init to match NF auxiliary attributes see MAE. 
(ii) `domain_map` same as MAE. 
(iii) `__call__` modified DILATE's forward

You can start from this example, available in the [pytorch losses notebook](https://github.com/Nixtla/neuralforecast/blob/main/nbs/losses.pytorch.ipynb):
```
class MAE(torch.nn.Module):
    """Mean Absolute Error

    Calculates Mean Absolute Error between
    `y` and `y_hat`. MAE measures the relative prediction
    accuracy of a forecasting method by calculating the
    deviation of the prediction and the true
    value at a given time and averages these devations
    over the length of the series.
    """    
    def __init__(self):
        super(MAE, self).__init__()
        self.outputsize_multiplier = 1
        self.output_names = ['']
        self.is_distribution_output = False

    def domain_map(self, y_hat: torch.Tensor):
        """
        Univariate loss operates in dimension [B,T,H]/[B,H]
        This changes the network's output from [B,H,1]->[B,H]
        """
        return y_hat.squeeze(-1)

    def __call__(self, y: torch.Tensor, y_hat: torch.Tensor,
                 mask: Union[torch.Tensor, None] = None):
        """
        **Parameters:**<br>
        `y`: tensor, Actual values.<br>
        `y_hat`: tensor, Predicted values.<br>
        `mask`: tensor, Specifies date stamps per serie to consider in loss.<br>

        **Returns:**<br>
        `mae`: tensor (single value).
        """
        if mask is None:
            mask = torch.ones_like(y)

        mae = torch.abs(y - y_hat) * mask
        mae = torch.mean(mae)
        return mae
```

**Reply by rnjv** [reply]
The slack channel complaints it's not open/require my email to be registered.
